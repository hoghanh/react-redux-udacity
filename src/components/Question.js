import React, { useEffect, useState } from 'react';
import { connect } from 'react-redux';
import { useParams } from 'react-router-dom';
import {
  Card,
  CardContent,
  CardHeader,
  Avatar,
  Typography,
  Button,
  FormControlLabel,
  FormControl,
  RadioGroup,
  Radio,
  Box,
  LinearProgress,
} from '@mui/material';

import Header from './Header';
import PageNotFound from './PageNotFound';
import { saveAnswer } from '../actions/shared';

const Question = (props) => {
  const { id } = useParams();
  const question = props.questions[id];

  const [value, setValue] = useState(0);
  const [optionOne, setOptionOne] = useState();
  const [optionTwo, setOptionTwo] = useState();
  const [totalVotes, setTotalVotes] = useState(0);
  const [isOptionOne, setIsOptionOne] = useState(false);

  useEffect(() => {
    if (question) {
      setOptionOne(question.optionOne);
      setOptionTwo(question.optionTwo);
      setTotalVotes(
        question.optionOne.votes.length + question.optionTwo.votes.length
      );
      setIsOptionOne(question.optionOne.votes.includes(props.authUser));
    }
  });

  const questionAnswered = Object.keys(
    props.users[props.authUser]?.answers
  ).includes(question?.id);

  const handleChange = (event) => {
    setValue(event.target.value);
  };

  const submitAnswer = (event) => {
    props.dispatch(saveAnswer(props.authUser, id, value));
  };

  const calculatePercentage = (voteAmount) =>
    Number(((voteAmount / totalVotes) * 100).toFixed(0));

  if (!question) {
    return <PageNotFound />;
  }

  return (
    <>
      <Header />
      <div style={{ display: 'flex', justifyContent: 'center' }}>
        <Card sx={{ width: 500, margin: 3, borderTop: '3px solid blue' }}>
          <CardHeader
            avatar={
              <Avatar
                src={props.users[question.author].avatarURL}
                alt='user-avatar'
              />
            }
            title={`${props.users[question.author].name} asked:`}
            style={{ backgroundColor: 'lightblue' }}
          />
          {questionAnswered ? (
            <CardContent>
              <Typography variant='h6'>
                <b>Your choice is:</b>
              </Typography>

              <Box
                sx={{
                  width: 400,
                  height: 75,
                  borderRadius: 1,
                  margin: '10px auto',
                  padding: 2,
                  backgroundColor: isOptionOne ? 'lightgreen' : 'mistyrose',
                }}
              >
                <Typography style={{ textAlign: 'center' }}>
                  {question.optionOne.text}
                </Typography>
                <Typography
                  variant='h6'
                  style={{ textAlign: 'center', color: 'darkblue' }}
                >{`${optionOne.votes.length} out of ${totalVotes} votes`}</Typography>
                <LinearProgress
                  variant='determinate'
                  value={calculatePercentage(optionOne.votes.length)}
                />
                {isOptionOne && (
                  <Typography style={{ textAlign: 'center' }}>
                    -------------- Your Vote --------------
                  </Typography>
                )}
              </Box>
              <Box
                sx={{
                  width: 400,
                  height: 75,
                  borderRadius: 1,
                  margin: '10px auto',
                  padding: 2,
                  backgroundColor: !isOptionOne ? 'lightgreen' : 'mistyrose',
                }}
              >
                <Typography style={{ textAlign: 'center' }}>
                  {question.optionTwo.text}
                </Typography>

                <Typography
                  variant='h6'
                  style={{ textAlign: 'center', color: 'darkblue' }}
                >{`${optionTwo.votes.length} out of ${totalVotes} votes`}</Typography>
                <LinearProgress
                  variant='determinate'
                  value={calculatePercentage(optionTwo.votes.length)}
                />
                {!isOptionOne && (
                  <Typography style={{ textAlign: 'center' }}>
                    -------------- Your Vote --------------
                  </Typography>
                )}
              </Box>
            </CardContent>
          ) : (
            <CardContent>
              <Typography variant='h6'>
                <b>Your choice is:</b>
              </Typography>
              <FormControl>
                <RadioGroup
                  name='controlled-radio-buttons-group'
                  value={value}
                  onChange={handleChange}
                >
                  <FormControlLabel
                    value='optionOne'
                    control={<Radio />}
                    label={question.optionOne.text}
                  />
                  <FormControlLabel
                    value='optionTwo'
                    control={<Radio />}
                    label={question.optionTwo.text}
                  />
                </RadioGroup>
              </FormControl>
              <div
                style={{
                  display: 'flex',
                  justifyContent: 'center',
                  marginTop: '10px',
                }}
              >
                <Button
                  variant='contained'
                  onClick={submitAnswer}
                  disabled={!value}
                >
                  Submit
                </Button>
              </div>
            </CardContent>
          )}
        </Card>
      </div>
    </>
  );
};

const mapStateToProps = (state) => ({
  authUser: state.authUser,
  questions: state.questions,
  users: state.users,
});

export default connect(mapStateToProps)(Question);
