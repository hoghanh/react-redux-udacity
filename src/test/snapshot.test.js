import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { toBeInTheDocument } from '@testing-library/jest-dom';
import { BrowserRouter } from 'react-router-dom';
import { createStore } from 'redux';
import renderer from 'react-test-renderer';
import Question from '../components/Question';
import { _getQuestions, _saveQuestion } from '../asset/_DATA';

const mockUsers = {
  sarahedo: {
    id: 'sarahedo',
    password: 'password123',
    name: 'Sarah Edo',
    avatarURL:
      'https://i.pinimg.com/originals/80/29/58/8029583119192523adf5396e0637cfbe.jpg',
    answers: {
      '8xf0y6ziyjabvozdd253nd': 'optionOne',
      '6ni6ok3ym7mf1p33lnez': 'optionOne',
      am8ehyc8byjqgar0jgpub9: 'optionTwo',
      loxhs1bqm25b708cmbf3g: 'optionTwo',
    },
    questions: ['8xf0y6ziyjabvozdd253nd', 'am8ehyc8byjqgar0jgpub9'],
  },
  mtsamis: {
    id: 'mtsamis',
    password: 'xyz123',
    name: 'Mike Tsamis',
    avatarURL:
      'https://i.pinimg.com/originals/01/b2/86/01b286554c4a16f9e9d67d432dc95e8d.jpg',
    answers: {
      xj352vofupe1dqz9emx13r: 'optionOne',
      vthrdm985a262al8qx3do: 'optionTwo',
      '6ni6ok3ym7mf1p33lnez': 'optionOne',
    },
    questions: ['6ni6ok3ym7mf1p33lnez', 'xj352vofupe1dqz9emx13r'],
  },
  zoshikanlu: {
    id: 'zoshikanlu',
    password: 'pass246',
    name: 'Zenobia Oshikanlu',
    avatarURL:
      'https://i.pinimg.com/originals/a5/09/2e/a5092e5e1f9d098d6a61f1dd1fc49454.jpg',
    answers: {
      xj352vofupe1dqz9emx13r: 'optionOne',
    },
    questions: [],
  },
};

const mockQuestions = {
  am8ehyc8byjqgar0jgpub9: {
    id: 'am8ehyc8byjqgar0jgpub9',
    author: 'sarahedo',
    timestamp: 1488579767190,
    optionOne: {
      votes: [],
      text: 'conduct a release retrospective 1 week after a release',
    },
    optionTwo: {
      votes: ['sarahedo'],
      text: 'conduct release retrospectives quarterly',
    },
  },
  loxhs1bqm25b708cmbf3g: {
    id: 'loxhs1bqm25b708cmbf3g',
    author: 'tylermcginnis',
    timestamp: 1482579767190,
    optionOne: {
      votes: [],
      text: 'have code reviews conducted by peers',
    },
    optionTwo: {
      votes: ['sarahedo'],
      text: 'have code reviews conducted by managers',
    },
  },
  vthrdm985a262al8qx3do: {
    id: 'vthrdm985a262al8qx3do',
    author: 'tylermcginnis',
    timestamp: 1489579767190,
    optionOne: {
      votes: ['tylermcginnis'],
      text: 'take a course on ReactJS',
    },
    optionTwo: {
      votes: ['mtsamis'],
      text: 'take a course on unit testing with Jest',
    },
  },
};

const mockState = {
  authUser: 'mtsamis',
  questions: mockQuestions,
  users: mockUsers,
};

const store = createStore((state = mockState, action) => state, {
  authUser: 'mtsamis',
  questions: mockQuestions,
  users: mockUsers,
});

jest.mock('react-redux', () => ({
  connect: jest.fn(() => (Component) => Component),
}));

test('_saveQuestion creates and saves a new question (snapshot)', async () => {
  const newQuestion = {
    optionOneText: 'Option A',
    optionTwoText: 'Option B',
    author: 'sarahedo',
  };

  const savedQuestion = await _saveQuestion(newQuestion);

  const questionTree = renderer
    .create(
      <BrowserRouter>
        <Question question={savedQuestion} {...mockState} />
      </BrowserRouter>
    )
    .toJSON();
  expect(questionTree).toMatchSnapshot();
});
