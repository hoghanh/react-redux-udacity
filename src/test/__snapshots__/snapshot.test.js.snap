// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_saveQuestion creates and saves a new question (snapshot) 1`] = `
<div
  className="MuiGrid-root MuiGrid-container css-go2vvz-MuiGrid-root"
>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-2 css-63b4os-MuiGrid-root"
    sm-offset={2}
  >
    <a
      href="/"
      onClick={[Function]}
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-sghohy-MuiButtonBase-root-MuiButton-root"
        disabled={false}
        onBlur={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        Back to home
      </button>
    </a>
  </div>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-8 css-o4bg9w-MuiGrid-root"
  >
    <img
      alt="Page Not Found"
      loading="lazy"
      src="404.png"
      srcSet="404.png"
      style={
        Object {
          "maxHeight": "100vh",
        }
      }
    />
  </div>
</div>
`;
